src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -303,33 +298,28 @@ private DecisionTree AddByType(DecisionTree.ByType byType, TypeSymbol type, Deci
                     }
                 }
             }
-            foreach (var kvp in byType.TypeAndDecision)
+
+            // if the last type is the type we need, add to it
+            DecisionTree result = null;
+            if (byType.TypeAndDecision.Count != 0)
             {
-                var MatchedType = kvp.Key;
-                var Decision = kvp.Value;
-                // See if matching Type matches this value
-                switch (ExpressionOfTypeMatchesPatternType(type, MatchedType, ref _useSiteDiagnostics))
+                var lastTypeAndDecision = byType.TypeAndDecision.Last();
+                if (lastTypeAndDecision.Key.TupleUnderlyingTypeOrSelf() == type.TupleUnderlyingTypeOrSelf())
```

Comment:
Filed #17088 for differences in nested tuple names.
Filed #17089 for the treatment of dynamic.

---
In reply to: [100601157](https://github.com/dotnet/roslyn/pull/16928#discussion_r100601157) [](ancestors = 100601157,100591895)